etf.py

```diff
@@ -0,0 +1,189 @@
+import pandas as pd
+import numpy as np
+import pyfolio as pf
+from typing import Optional, List, Dict
+from decimal import Decimal
+import model
+from dataclasses import asdict
+from ib_insync import util
+import ib_insync as ibapi
+
+
+def etf(portfolio: pd.DataFrame) -> pd.Series:
+    """
+    Returns a time series representing the investment of $1 in a basket of instruments weighted proportionally by the given weights.
+
+    @param val: A DataFrame of instruments containing Open, Close, and Weight data indexed by Date.
+    """
+    # Initialize a zero'd out T-sized array where T is the length of the date range.
+    etf = np.zeros(portfolio.loc['open'].shape[0])
+    etf[0] = Decimal(1)  # Initial AUM for this instrument is $1.
+
+    # Initialize an I x T array of holdings over time so we can keep track of how we've divided our AUM
+    # across the given instruments over time.
+    hodls = np.zeros(portfolio.loc['open'].shape)
+
+    for t in range(1, etf.shape[0]):
+        portfolio_sum: Decimal = Decimal(0)
+
+        for i in portfolio.columns:
+            # Calculate the absolute change of asset i from time t to t - 1.
+            change = delta(portfolio, i, t)
+
+            # TODO: Figure out where to get dividend information if it's not calculated into the price.
+            dividend = Decimal(0)
+            exchange_rate = Decimal(1)
+
+            # Calculate the holdings of instrument i at t - 1 and store it in our hodls matrix.
+            holdings_t_minus_1 = holdings(portfolio, hodls, i, t - 1,
+                                          etf[t - 1])
+            hodls[t - 1][portfolio.columns.get_loc(i)] = holdings_t_minus_1
+
+            portfolio_sum += holdings_t_minus_1 * exchange_rate * (change +
+                                                                   dividend)
+
+        # The AUM of the etf at time t is the AUM of the etf at t - 1 plus
+        # the sum of the changes in the instruments from t-1 to t accounting
+        # for exchange rates, transaction costs, and dividends.
+        etf[t] = Decimal(etf[t - 1]) + portfolio_sum
+    return etf
+
+
+def portfolio_to_returns(portfolio: pd.DataFrame, timezone: str) -> pd.Series:
+    index = portfolio.index.levels[1].tz_localize(timezone)
+    prices = pd.Series(etf(portfolio), index=index)
+    return prices_to_daily_returns(prices)
+
+
+def prices_to_daily_returns(prices: pd.Series) -> pd.Series:
+    """
+    Calculates daily returns for a Series of prices.
+    Note: drops the first value of the given Series.
+    """
+    return (prices / prices.shift(1) - 1)[1:]
+
+
+def positions_to_dataframe(positions: List[model.Position]) -> pd.DataFrame:
+    """
+    Returns a dataframe of positions with an additional `value` and `allocation` columns
+    calculated from the averagePrice and quantity.
+    """
+    is_stock = lambda position: type(position.instrument) == model.Stock
+    stocks = list(map(lambda p: asdict(p), filter(is_stock, positions)))
```

Comment:
TIOLI but I mostly find list comprehensions more readable:

```suggestion
    stocks = [asdict(p) for p in positions if type(position.instrument) == model.Stock]
```